README.md
[![Build Status](https://travis-ci.org/railslove/iban_calculator.svg?branch=master)](https://travis-ci.org/railslove/iban_calculator)

# IbanCalculator

A wrapper for ibanrechner.de API. It allows converting bank account data from legacy syntax to new SEPA IBAN/BIC.

## Installation

Add this line to your application's Gemfile:

    gem 'iban_calculator'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install iban_calculator

## Usage

In order to use iban calculator, you need to create an account at [iban-bic.com](http://www.iban-bic.com/).

In case you are using rails, configure your app like this:

    # config/initializers/iban_calculator.rb
    IbanCalculator.user = 'your_username',
    IbanCalculator.password = 'your_password'
    IbanCalculator.logger = Rails.logger

Whenever you need to convert European legacy account data to new SEPA IBAN format:

    # app/models/your_model.rb
    IbanCalculator.calculate_iban country: 'DE', account_number: '123', bank_code: '456'

Example data can be found at: http://www.iban-bic.com/sample_accounts.html

You can also validate a given IBAN and fetch additional data about it:

    IbanCalculator.validate_iban 'AL90208110080000001039531801'


## Contributing

1. Fork it ( https://github.com/[my-github-username]/iban_calculator/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
0.2.1
-----
* make it work with newer version of active_support too

0.2.0
-----
* [lower active_support dependency](https://github.com/railslove/iban_calculator/pull/7)

0.1.1
-----

* added missing failure code for iban checksum
* fixed error condition, so not every request is concidered an error ;)
* `#as_json` will have stringified keys
* `#valdiate_iban` and `#calculate_iban` should respond with mostly the same format
* rescue from `ArgumentError` happening when the date is present but not valid

0.1.0
-----

* Allow invalid IBANs to bubble though and not to fail spectacular

0.0.3
-----

* Raise an exception when an unexpected response is returned (i.e. invalid user and password)

0.0.2
-----

* Add support for validating IBANs and fetching related data such as legacy data

0.0.1
-----

* Initial release with support for calculating IBANs based on legacy data


spec/spec_helper.rb
require 'iban_calculator'

RSpec.configure do |config|
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  if config.files_to_run.one?
    config.default_formatter = 'doc'
  end

  config.profile_examples = 3

  config.order = :random

  Kernel.srand config.seed

  config.expect_with :rspec do |expectations|
    expectations.syntax = :expect
  end

  config.mock_with :rspec do |mocks|
    mocks.syntax = :expect
    mocks.verify_partial_doubles = true
  end
end


spec/iban_bic_spec.rb
require 'ostruct'

describe IbanCalculator::IbanBic do
  let(:config) {
    OpenStruct.new(
      user: 'user',
      password: 'pass',
      url: 'url',
      logger: Logger.new(STDOUT),
      read_timeout: 5,
      open_timeout: 5
    )
  }

  subject { described_class.new(config) }

  before { allow(subject.logger).to receive(:info) }

  describe '#italian_account_number' do
    it 'returns an empty hash if not all fields are provided' do
      expect(subject.italian_account_number).to eq({})
    end

    it 'returns hash with correct account number if valid data is provided' do
      expect(subject.italian_account_number(
        'country' => 'IT',
        'cab' => '03280',
        'abi' => '03002',
        'cin' => 'D',
        'account' => '400162854',
      )).to eq('account' => 'D0300203280000400162854')
    end
  end

  describe '#default_payload' do
    it 'includes account data' do
      expect(subject.default_payload).to match hash_including(user: 'user', password: 'pass')
    end
  end

  describe '#process_bic_candidates' do
    context 'known single BIC payload' do
      let(:payload) { valid_payload[:bic_candidates] }

      it 'returns an array' do
        expect(subject.process_bic_candidates(payload)).to be_kind_of(Array)
      end

      it 'returns its bank\'s bic' do
        expect(subject.process_bic_candidates(payload).first).to match hash_including(bic: 'BYLADEM1001')
      end

      it 'returns its bank\'s zip' do
        expect(subject.process_bic_candidates(payload).first).to match hash_including(zip: '10117')
      end

      it 'returns its bank\'s city' do
        expect(subject.process_bic_candidates(payload).first).to match hash_including(city: 'Berlin')
      end

      context 'empty fields' do
        let(:payload) { {:item=>{
                            :bic=>"UNCRITMM",
                            :zip=>{:"@xsi:type"=>"xsd:string"},
                            :city=>{:"@xsi:type"=>"xsd:string"},
                            :wwwcount=>"0",
                            :sampleurl=>{:"@xsi:type"=>"xsd:string"},
                            :"@xsi:type"=>"tns:BICStruct"
                          },
                          :"@xsi:type"=>"SOAP-ENC:Array",
                          :"@soap_enc:array_type"=>"tns:BICStruct[1]"} }

        it 'ignores empty zip' do
          expect(subject.process_bic_candidates(payload).first.keys).to_not include(:zip)
        end

        it 'ignores empty city' do
          expect(subject.process_bic_candidates(payload).first.keys).to_not include(:city)
        end
      end
    end

    context 'unknown payload' do
      let(:payload) { { :items => [], :"@xsi:type" => 'SOAP-ENC:Array' } }

      it 'logs the payload' do
        subject.process_bic_candidates(payload) rescue
        expect(subject.logger).to have_received(:info)
      end

      it 'raises an exception' do
        expect { subject.process_bic_candidates(payload) }.to raise_exception(ArgumentError)
      end
    end
  end

  describe '#formatted_result' do
    before { allow(subject).to receive(:process_bic_candidates).and_return(['data']) }

    it 'returns a valid ruby date for last update date' do
      expect(subject.formatted_result(valid_payload)[:updated_at]).to be_kind_of(Date)
    end

    it 'transforms the list of bic candidates' do
      subject.formatted_result(valid_payload)
      expect(subject).to have_received(:process_bic_candidates)
    end

    it 'includes iban' do
      expect(subject.formatted_result(valid_payload).keys).to include(:iban)
    end
  end

  describe '#iban_payload' do
    context 'italian data is provided' do
      before { allow(subject).to receive(:italian_account_number).and_return({ 'account' => 'italy-123' }) }

      it 'normalizes italian account data' do
        subject.iban_payload({})
        expect(subject).to have_received(:italian_account_number)
      end

      it 'merges italian data' do
        expect(subject.iban_payload({ 'country' => 'IT' })).to match(hash_including(account: 'italy-123'))
      end

      it 'strips italian data' do
        expect(subject.iban_payload({ 'cin' => '123' }).keys).to_not include('cin')
      end
    end

    it 'adds default payload' do
      expect(subject.iban_payload({}).keys).to include(:user, :password, :legacy_mode)
    end

    it 'overrides default data' do
      expect(subject.iban_payload({ bank_code: '123' })).to match hash_including(bank_code: '123')
    end

    it 'replaces account_number with account' do
      expect(subject.iban_payload({ account_number: '123' })).to match hash_including(account: '123')
    end
  end

  describe '#calculate_iban' do
    before { allow(subject.client).to receive(:call).and_return(response) }

    context 'valid response' do
      let(:response) { double(body: { calculate_iban_response: { return: valid_payload } }) }

      it 'returns a formatted response' do
        allow(subject).to receive(:formatted_result)
        subject.calculate_iban({})
        expect(subject).to have_received(:formatted_result)
      end

      it 'calls the client with the generated payload' do
        subject.calculate_iban({})
        expect(subject.client).to have_received(:call).with(:calculate_iban, message: anything)
      end
    end

    context 'probably valid response' do
      let(:response) { double(body: { calculate_iban_response: { return: valid_payload.merge(return_code: '32') } }) }

      it 'logs a message' do
        subject.calculate_iban({}) rescue IbanCalculator::InvalidData
        expect(subject.logger).to have_received(:info).with(/needs manual check/)
      end

      it 'returns a formatted response' do
        allow(subject).to receive(:formatted_result)
        subject.calculate_iban({})
        expect(subject).to have_received(:formatted_result)
      end
    end

    context 'invalid response' do
      let(:response) { double(body: { calculate_iban_response: { return: valid_payload.merge(return_code: '128') } }) }

      it 'logs a message' do
        subject.calculate_iban({}) rescue IbanCalculator::InvalidData
        expect(subject.logger).to have_received(:info).with(/iban check invalid/)
      end

      it 'fails with invalid data exception' do
        expect{ subject.calculate_iban({}) }.to raise_exception(IbanCalculator::InvalidData)
      end
    end

    context 'server error response' do
      let(:response) { double(body: { calculate_iban_response: { return: valid_payload.merge(return_code: '65536') } }) }

      it 'logs a message' do
        subject.calculate_iban({}) rescue IbanCalculator::ServiceError
        expect(subject.logger).to have_received(:info).with(/iban check failed/)
      end

      it 'fails with service error exception' do
        expect{ subject.calculate_iban({}) }.to raise_exception(IbanCalculator::ServiceError)
      end
    end
  end

  def valid_payload
    {
      :iban=>"DE59120300001111236988",
      :result=>"passed",
      :return_code=>"0",
      :ibanrueck_return_code=>{
        :"@xsi:type"=>"xsd:string"
      },
      :checks=>{
        :item=>["length", "bank_code", "account_number"],
        :"@xsi:type"=>"SOAP-ENC:Array",
        :"@soap_enc:array_type"=>"xsd:string[3]"
      },
      :bic_candidates=>{
        :item=>{
          :bic=>"BYLADEM1001",
          :zip=>"10117",
          :city=>"Berlin",
          :wwwcount=>"0",
          :sampleurl=>{:"@xsi:type"=>"xsd:string"},
          :"@xsi:type"=>"tns:BICStruct"
        },
        :"@xsi:type"=>"SOAP-ENC:Array",
        :"@soap_enc:array_type"=>"tns:BICStruct[1]"
      },
      :country=>"DE",
      :bank_code=>"12030000",
      :alternative_bank_code=>{:"@xsi:type"=>"xsd:string"},
      :bank=>"Deutsche Kreditbank Berlin",
      :bank_address=>{:"@xsi:type"=>"xsd:string"},
      :bank_url=>{:"@xsi:type"=>"xsd:string"},
      :branch=>{:"@xsi:type"=>"xsd:string"},
      :branch_code=>{:"@xsi:type"=>"xsd:string"},
      :in_scl_directory=>"yes",
      :sct=>"yes",
      :sdd=>"yes",
      :b2b=>"yes",
      :account_number=>"1011856976",
      :alternative_account_number=>{:"@xsi:type"=>"xsd:string"},
      :account_validation_method=>"00",
      :account_validation=>"Methode 00, Konto 1111236988, BLZ 12030000, Prüfziffer 6 steht an Position 10, erwartete Prüfziffer: 6. Überblick über die Berechnung: Nimm die Ziffern auf den Positionen 1 bis 9 - hier: 1111236988 -, multipliziere sie von rechts nach links mit den Gewichten 2,1,2,1,2,1,2,1,2, addiere die Quersummen der Produkte, bilde den Rest der Division durch 10, ziehe das Ergebnis von 10 ab,  und das Ergebnis modulo 10 ist die erwartete Prüfziffer.",
      :length_check=>"passed",
      :account_check=>"passed",
      :bank_code_check=>"passed",
      :bic_plausibility_check=>{:"@xsi:type"=>"xsd:string"},
      :data_age=>"20140525",
      :iba_nformat=>"DEkk BBBB BBBB CCCC CCCC CC",
      :formatcomment=>"B = sort code (BLZ), C = account No.",
      :balance=>"4",
      :"@xsi:type"=>"tns:IBANCalcResStruct"
    }
  end
end


spec/invalid_data_spec.rb
describe IbanCalculator::InvalidData do
  subject { described_class.new('', 128) }

  describe '#resolve_error_code' do
    it 'handles single errors' do
      expect(subject.resolve_error_code(128)).to eq(account_number: [:checksum_failed])
    end

    it 'can combine multiple errors' do
      expect(subject.resolve_error_code(128 + 256)).to eq(account_number: [:checksum_failed], bank_code: [:not_found])
    end

    it 'can combine multiple errors of same attribute' do
      expect(subject.resolve_error_code(128 + 512)).to eq(account_number: [:checksum_failed, :invalid_length])
    end

    it 'can combine multiple errors of same attribute' do
      expect(subject.resolve_error_code(512)).to eq(account_number: [:invalid_length])
    end
  end

  describe '#error_codes' do
    it 'ignores zeros' do
      expect(subject.error_codes(0)).to eq([])
    end

    it 'detects unique values' do
      expect(subject.error_codes(1)).to eq([1])
    end

    it 'detects compounded values' do
      expect(subject.error_codes(3)).to eq([1, 2])
    end

    it 'detects 512' do
      expect(subject.error_codes(512)).to eq([512])
    end

    it 'detects compounded values up to 8192' do
      expect(subject.error_codes(8193)).to eq([1, 8192])
    end
  end

  describe '#known_error_codes' do
    it 'detects 512' do
      expect(subject.known_error_codes(512)).to eq([512])
    end
  end

  describe '#error' do
    it 'sets error on object' do
      expect(described_class.new('test', 512).errors).to eq(account_number: [:invalid_length])
    end
  end
end


lib/iban_calculator.rb
require 'iban_calculator/version'
require 'active_support/version'
require 'active_support/configurable'
require 'active_support/core_ext/hash'
require 'logger'

require 'iban_calculator/bank'
require 'iban_calculator/bic_candidate'
require 'iban_calculator/iban_bic'
require 'iban_calculator/iban_validator_response'
require 'iban_calculator/invalid_data'
require 'iban_calculator/active_support' if ActiveSupport::VERSION::MAJOR == 3

module IbanCalculator
  # Extensions
  include ActiveSupport::Configurable

  # Configuration
  config_accessor(:url)          { 'https://ssl.ibanrechner.de/soap/?wsdl' }
  config_accessor(:user)         { '' }
  config_accessor(:password)     { '' }
  config_accessor(:logger)       { Logger.new(STDOUT) }
  config_accessor(:read_timeout) { 5 }
  config_accessor(:open_timeout) { 5 }

  # Errors
  ServiceError = Class.new(StandardError)

  def self.calculate_iban(attributes = {})
    client = IbanBic.new(config)
    client.calculate_iban(attributes)
  end

  def self.validate_iban(iban)
    response = execute(:validate_iban, iban: iban, user: config.user, password: config.password)
    IbanValidatorResponse.new(response.body[:validate_iban_response][:return])
  end

  def self.execute(method, options = {})
    client = Savon.client(wsdl: config.url, logger: config.logger, read_timeout: config.read_timeout, open_timeout: config.open_timeout)
    client.call(method, message: options).tap do |response|
      status = response.body[:"#{method}_response"][:return][:result]
      fail(ServiceError, status) unless response.body[:"#{method}_response"][:return][:return_code]
    end
  end
end


lib/iban_calculator/invalid_data.rb
module IbanCalculator
  class InvalidData < StandardError
    CODES = {
      128  => [:account_number, [:checksum_failed]],
      256  => [:bank_code,      [:not_found]],
      512  => [:account_number, [:invalid_length]],
      1024 => [:bank_code,      [:invalid_length]],
      2048 => [:iban,           [:checksum_failed]],
      4096 => [:base,           [:data_missing]],
      8192 => [:country,        [:not_supported]],
    }

    attr_accessor :errors

    def initialize(msg, error_code)
      self.errors = resolve_error_code(error_code)
      super(msg)
    end

    def resolve_error_code(error_code)
      known_error_codes(error_code).reduce(Hash.new([])) do |hsh, item|
        error = CODES[item]
        hsh[error[0]] += error[1]
        hsh
      end
    end

    def known_error_codes(error_code)
      error_codes(error_code) & CODES.keys
    end

    def error_codes(n)
      (0..13).map { |i| n & 2**i } - [0]
    end
  end
end


lib/iban_calculator/bic_candidate.rb
module IbanCalculator
  class BicCandidate
    attr_accessor :bic, :city, :sample_url, :www_count, :zip

    def self.build_list(raw_response)
      Array.wrap(raw_response[:item]).map { |candidate| new(candidate) }
    end

    def initialize(raw_attributes = {})
      self.bic = raw_attributes[:bic]
      self.zip = string_or_default(raw_attributes[:zip])
      self.city = string_or_default(raw_attributes[:city])
      self.sample_url = string_or_default(raw_attributes[:sampleurl])
      self.www_count = raw_attributes[:wwwcount].to_i
    end

    def source
      www_count > 0 ? :www : :directory
    end

    def as_json(opts = {})
      {
        bic: bic,
        zip: zip,
        city: city,
        sample_url: sample_url,
        www_count: www_count,
      }.deep_stringify_keys!
    end

    private

    def string_or_default(input, default = '')
      input.kind_of?(String) ? input : default
    end
  end
end


lib/iban_calculator/active_support.rb
# these are methods implemented in active_support 4 which
# are required to run this gem with active_support 3

module IbanCalculator
  include ActiveSupport::Configurable

  # taken from: activesupport/lib/active_support/core_ext/hash/keys.rb
  Hash.class_eval do
    def deep_stringify_keys!
      deep_transform_keys!{ |key| key.to_s }
    end

    def deep_transform_keys!(&block)
      keys.each do |key|
        value = delete(key)
        self[yield(key)] = value.is_a?(Hash) ? value.deep_transform_keys!(&block) : value
      end
      self
    end
  end
end

class << IbanCalculator
  alias_method :old_config_accessor, :config_accessor

  def config_accessor(*names)
    old_config_accessor(*names)

    return unless block_given?

    names.each do |name|
      send("#{name}=", yield)
    end
  end
end


lib/iban_calculator/iban_validator_response.rb
module IbanCalculator
  class IbanValidatorResponse
    CHECKS = {
      length: :length_check,
      account_number: :account_check,
      bank_code: :bank_code_check,
      iban_checksum: :iban_checksum_check,
    }

    attr_accessor :raw_response

    def initialize(raw_response)
      self.raw_response = raw_response
    end

    def valid?
      return_code < 128
    end

    def return_code
      @return_code ||= raw_response[:return_code].to_i
    end

    def bic_candidates
      @bic_candidates ||= BicCandidate.build_list(raw_response[:bic_candidates])
    end

    def bank
      @bank ||= begin Bank.new({
          code: string_or_default(raw_response[:bank_code]),
          name: string_or_default(raw_response[:bank]),
          country: string_or_default(raw_response[:country]),
          address: string_or_default(raw_response[:bank_address]).strip,
          url: string_or_default(raw_response[:bank_url]),
          branch: string_or_default(raw_response[:branch]),
          branch_code: string_or_default(raw_response[:branch_code]),
        })
      end
    end

    def account_number
      @account_number ||= string_or_default(raw_response[:account_number], nil)
    end

    def checks
      CHECKS.each_with_object({}) do |(app_key, api_key), result|
        result[app_key] = string_or_default(raw_response[api_key], 'not_checked')
      end
    end

    def updated_at
      @data_created_at ||= Date.parse(raw_response[:data_age]) if string_or_default(raw_response[:data_age], nil)
    rescue ArgumentError
      nil
    end

    def errors
      @errors ||= InvalidData.new('', return_code).errors
    end

    def as_json(opts = {})
      {
        valid: valid?,
        errors: errors,
        account_number: account_number,
        bank: bank.name,
        bank_code: bank.code,
        bics: bic_candidates.map { |c| c.as_json(opts) },
        updated_at: updated_at,
        checks: checks,
      }.deep_stringify_keys!
    end

    private

    def string_or_default(input, default = '')
      input.kind_of?(String) ? input : default
    end
  end
end


lib/iban_calculator/iban_bic.rb
require 'savon'

# Return codes and their meaning:
#
# 0 = all checks were successful
#
# 1 = sub account number has been added automatically
# 2 = account number did not include a checksum
# 4 = checksum has not been checked
# 8 = bank code has not been checked
#
# 32 = A sub account number might be required, but could not be determined autoamtically
#
# 128 = checksum for account_number is invalid
# 256 = bank_code could not be found is database
# 512 = account_number has an invalid length
# 1024 = bank_code has an invalid length
# 4096 = data is missing (i.e. country code)
# 8192= country is not yet supported
#
module IbanCalculator
  class IbanBic
    ITALIAN_IBAN_LENGTH = 27
    PREFIX_AND_CHECKSUM_LENGTH = 4

    VALID_RESPONSE_CODE = 0..31
    PROBABLY_VALID_RESPONSE_CODE = 32..127
    SERVICE_ERROR_RESPONSE_CODE = 65536

    attr_accessor :user, :password, :url, :logger, :config

    def initialize(config)
      self.user = config.user
      self.password = config.password
      self.url = config.url
      self.logger = config.logger
      self.config = config
    end

    # You should provide country, bank_code, and account_number. (cin, abi, and cab for Italian accounts)
    def calculate_iban(attributes)
      payload = iban_payload(attributes)
      response = client.call(:calculate_iban, message: payload).body[:calculate_iban_response][:return]
      log "iban lookup attributes=#{attributes} payload=#{payload} response=#{response}"

      case return_code = response[:return_code].to_i
      when VALID_RESPONSE_CODE
        formatted_result(response)
      when PROBABLY_VALID_RESPONSE_CODE
        log "iban check needs manual check return_code=#{return_code}"
        formatted_result(response)
      when SERVICE_ERROR_RESPONSE_CODE
        log "iban check failed return_code=#{return_code}"
        fail ServiceError, 'Service could not handle the request'
      else
        log "iban check invalid return_code=#{return_code}"
        fail InvalidData.new('Invalid input data', return_code)
      end
    end

    def italian_account_number(attributes = {})
      return {} unless attributes['country'].to_s.upcase == 'IT'
      left_length = ITALIAN_IBAN_LENGTH - PREFIX_AND_CHECKSUM_LENGTH - attributes['account'].length
      left_side = [attributes['cin'], attributes['abi'], attributes['cab']].join.ljust(left_length, '0')
      { 'account' => left_side + attributes['account'] }
    end

    def default_payload
      { country: '', bank_code: '', account: '', user: user, password: password, bic: '', legacy_mode: 0 }
    end

    def formatted_result(data)
      { iban: data[:iban],
        bics: process_bic_candidates(data[:bic_candidates]),
        country: data[:country],
        bank_code: data[:bank_code],
        bank: data[:bank],
        account_number: data[:account_number],
        updated_at: Date.parse(data[:data_age]) }
    end

    def process_bic_candidates(candidates)
      [candidates[:item].select { |key, value| [:bic, :zip, :city].include?(key) && value.kind_of?(String) }]
    rescue
      log "Could not handle candidates=#{candidates}"
      fail ArgumentError, "Could not handle BIC response"
    end

    def iban_payload(attributes)
      attributes = attributes.with_indifferent_access
      attributes['account'] = attributes.delete('account_number')
      normalized_attributes = attributes.merge(italian_account_number(attributes))
      payload = normalized_attributes.select { |k,_| %w(country account bank_code).include?(k) }
      default_payload.merge(payload.symbolize_keys)
    end

    def log(message)
      logger.info message
    end

    def client
      @client ||= Savon.client(wsdl: url, read_timeout: config.read_timeout, open_timeout: config.open_timeout)
    end
  end
end
